### DISCUSSION TITLE
Beta test the new features coming in Tailwind CSS v3.2

### DISCUSSION THREAD

**Discussion by adamwathan** [tailwindlabs/tailwindcss#9504] [Announcements]
Hey everyone! We're planning to release Tailwind CSS v3.2 in the next couple of weeks, and we'd love to get some more people flexing the new features we're working on before we tag it and can't turn back.

A lot of the new things we've been working on this time around involve brand new syntax and concepts. With Tailwind being such a widely used tool these days, we'd really love for people to put this stuff to test ahead of the release and make sure we aren't making any API decisions that create really painful integration problems when working with tools we don't use ourselves.

To install the latest pre-release, install our insiders build from npm:

```sh
npm install -D tailwindcss@insiders
```

You can also try the new features directly in [Tailwind Play](https://play.tailwindcss.com/TZxjYqVrKQ) by choosing "Insiders" from the version dropdown, but it's better for us if you actually install it in your project so we can get more information about how all this stuff is going to work alongside different UI libraries, templating languages, and build tools.

Here's a breakdown of the new stuff we'd love some help putting through its paces!

- [Multiple config files in one project using `@config`](#multiple-config-files-in-one-project-using-config)
- [New variant for `@supports` rules](#new-variant-for-supports-rules)
- [Dynamic `group-[...]` and `peer-[...]` variants](#dynamic-group-and-peer-variants)
- [Parameterized variants with `matchVariant`](#parameterized-variants-with-match-variant)
- [Nested `group` and multiple `peer` support using &lt;labels&gt;](#nested-group-and-multiple-peer-support-using-lt-labels-gt)

---

<a name="multiple-config-files-in-one-project-using-config"></a>

## Multiple config files in one project using `@config`

We've added a new `@config` directive that you can use in a CSS file to specify which Tailwind CSS config to use for that file:

```css
@config "./tailwind.admin.config.js"
@tailwind base;
@tailwind components;
@tailwind utilities;
```

This makes it a lot easier to build multiple stylesheets in a single project that have separate Tailwind configurations. For example, you might have one config file for the customer-facing part of your site, and another config for the admin/backend area.

Config files specified with `@config` are resolved relative to the CSS file where the `@config` directive is being used.

One known pain-point right now is that when using `postcss-import`, you can't put `@config` at the top of the file because `postcss-import` is strict about all `@import` statements having to come first as per the CSS spec. We might wrap `postcss-import` with our own import plugin to remove this limitation, but for now if you're using `@import` statements, your `@config` directive will need to come after them:

```css
@import "tailwindcss/base";
@import "tailwindcss/components";
@import "tailwindcss/utilities";
@config "./tailwind.admin.config.js";
```

The Tailwind CSS IntelliSense plugin also doesn't have support for this stuff in any sort of automatic way yet, so you might not see the completions you expect depending on how you have things configured. We're working on this though, and hope to have a solution in place before the actual v3.2 release.

---

<a name="new-variant-for-supports-rules"></a>

## New variant for `@supports` rules

You can now conditionally style things based on whether a certain feature is supported in the user's browser with the `supports-[...]` variant, which generates [`@supports rules`](https://developer.mozilla.org/en-US/docs/Web/CSS/@supports) under the hood.

```html
<div class="flex supports-[display:grid]:grid ...">
  <!-- ... -->
</div>
```

The `supports-[...]` variant takes anything you'd use with `@supports (...)` between the square brackets, like a property/value pair, and even expressions using `and` and `or`.

For terseness, if you only need to check if a property is supported (and not a specific value), you can just specify the property name:

```html
<div class="bg-black/75 supports-[backdrop-filter]:bg-black/25 supports-[backdrop-filter]:backdrop-blur ...">
  <!-- ... -->
</div>
```

---

<a name="dynamic-group-and-peer-variants"></a>

## Dynamic `group-[...]` and `peer-[...]` variants

We're making it possible to create custom `group-*` and `peer-*` variants on the fly by passing your selector between square brackets

```html
<div>
  <!-- ... -->
  <a class="peer" href="#">...</a>
  <a class="peer-[:local-link]:ml-3">
    <!-- ... -->
  </a>
  <!-- ... -->
</div>
```

Whatever you pass will just get tacked on to the end of the `.group` or `.peer` class in the selector by default, but if you need to do something more complex you can use the `&` character to mark where `.group`/`.peer` should end up in the final selector relative to the selector you are passing in

```html
<div class="group">
  <div class="group-[:nth-of-type(3)_&]:border-black">
    <!-- ... -->
  </div>
</div>
```

---

<a name="parameterized-variants-with-match-variant"></a>

## Parameterized variants with `matchVariant`

Both the new `supports-[..]` feature and dynamic `group-*`/`peer-*` variants are powered by a new `matchVariant` plugin API that you can use to create your own dynamic/parameterized variants.

Here's an example of creating a `placement-*` variant for some imaginary tooltip library that uses a `data-placement` attribute to tell you where the tooltip is currently positioned:

```js
let plugin = require("tailwindcss/plugin");

module.exports = {
  // ...
  plugins: [
    plugin(function ({ matchVariant }) {
      matchVariant(
        "placement",
        ({ value }) => {
          return `&[data-placement=${value}]`;
        },
        {
          t: "top",
          r: "right",
          b: "bottom",
          l: "left",
        }
      );
    }),
  ],
};
```

The variant defined above would give you variants like `placement-t` and `placement-b`, but would also support the arbitrary portion in square brackets, so if this imaginary tooltip library had other potential values that you didn't feel the need to create built-in values for, you could still do stuff like this:

```html
<div class="placement-[top-start]:mb-2 ...">
  <!-- ... -->
</div>
```

When defining a custom variant with this API, it's often important that you have some control over which order the CSS is generated in to make sure each class has the right precedence with respect to other values that come from the same variant. To support this, there's a `sort` function you can provide when defining your variant:

```js
matchVariant("min", ({ value }) => `@media (min-width: ${value})`, {
  sort(a, z) {
    return parseInt(a) - parseInt(z);
  },
});
```

---

<a name="nested-group-and-multiple-peer-support-using-lt-labels-gt"></a>

## Nested `group` and multiple `peer` support using &lt;labels&gt;

Sometimes you can run into problems when you have multiple `group` chunks nested within each other because Tailwind has no real way to disambiguate between them.

To solve this, we're trying out this new idea called _labels_, which are a new dynamic chunk that lives in the variant that you can use to give each group/peer your own identifier.

Here's what it looks like:

```html
<div class="group<sidebar> ...">
  <!-- ... -->
  <div class="group<navitem> ...">
    <a href="#" class="bg-black/25 group<sidebar>-hover:bg-black/50 group<navitem>-hover:bg-black/75">...</a>
  </div>
  <!-- ... -->
</div>
```

This lets you give each group a clear name that makes sense for that context on the fly, and Tailwind will generate the necessary CSS to make it work.

I'm really excited to have a solution out there for this because it's something I've been trying to land on a good approach for solving for several years, and this is the first thing we've come up with that really feels like it offers the power and flexibility I think it should.

This feature is probably the riskiest one in terms of creating issues with different tools though, because we're using new characters in the class names (`<` and `>`). So really flex this one and see how it works for you in your projects with the other libraries and build tools you like to use.

---

So that's everything! We have a few other smaller ideas we might try to squeeze in before now and the actual release, but this is all of the stuff I'd really appreciate help testing so we can make any API changes we need to make before it's too late and we become prisoners to the cruel bastard that is semantic versioning.

So install `tailwindcss@insiders`, beat the hell out of it, and let us know what you find!

**Comment by didier** [comment]
Though I understand the use-case, I can't possibly imagine writing `<a href="#" class="bg-black/25 group<sidebar>-hover:bg-black/50 group<navitem>-hover:bg-black/75">...</a>` without wanting to throw my laptop at a wall...

**Reply by colinaut** [reply]
Yeah just visually it feels like bad DX to use <> within tags, even if it doesn't mess with prettier or the like. Why not use `group(navbar)`?

**Reply by didier** [reply]
Looks much better, yeah. Wonder if CSS allows for it in classnames?

**Reply by adamwathan** [reply]
Considering using this syntax instead:

```html
<div class="group/sidebar ...">
  <!-- ... -->
  <div class="group/navitem ...">
    <a href="#" class="bg-black/25 group-hover/sidebar:bg-black/50 group-hover/navitem:bg-black/75">...</a>
  </div>
  <!-- ... -->
</div>
```

This way it's more like the opacity modifier syntax we use for utilities and not so much a brand new syntax/concept — it's just the idea of a trailing modifier applied to variants just like it can apply to utilities.

Right now I'm like 80% sure we'll make this change, always felt a little off to me introducing "labels" as a first-class concept in the framework if there was a more general purpose solution, and I think "modifiers" like we already use for opacity stuff is more universally useful. For sure there will be other situations where some sort of ad-hoc extra data like this will be useful 👍

**Reply by mryechkin** [reply]
Agreed with others - using `<>` within tags feels gross and unreadable. The slash makes much more sense considering it's syntax that's already used :+1:

**Reply by IRediTOTO** [reply]
How about group[navitem]-hover:bg-black/75

**Reply by adamwathan** [reply]
We've considered both `group(navitem)-hover:...` and `group[navitem]-hover:...` but not really excited about either option. The parentheses I think just looks ugly for this purpose, I have this pipe dream of one day figuring out how to make all this stuff composable so you can do `group(hover):underline` instead of `group-hover` and have it magically work with even custom variants created by end users, so I don't want to "spend" that syntax on this labels feature just in case we ever crack that.

The square brackets isn't terrible but it just feels weird to use them again but for a different reason than we currently use them.

**Reply by hussain-attari** [reply]
> Considering using this syntax instead:
> 
> ```
> <div class="group/sidebar ...">
>   
>   <div class="group/navitem ...">
>     <a href="#" class="bg-black/25 group-hover/sidebar:bg-black/50 group-hover/navitem:bg-black/75">...</a>
>   </div>
>   
> </div>
> ```
> 
> This way it's more like the opacity modifier syntax we use for utilities and not so much a brand new syntax/concept — it's just the idea of a trailing modifier applied to variants just like it can apply to utilities.
> 
> Right now I'm like 80% sure we'll make this change, always felt a little off to me introducing "labels" as a first-class concept in the framework if there was a more general purpose solution, and I think "modifiers" like we already use for opacity stuff is more universally useful. For sure there will be other situations where some sort of ad-hoc extra data like this will be useful 👍

How about using `group.sidebar`

It also means the group option is getting concatenated with a div element namely sidebar.

We can then have `group.sidebar:hover`

**Reply by brandonmcconnell** [reply]
Personally, I feel like using a dot `.` feels to similar to a plain CSS class name which this is not, so I’d want to avoid that.

@adamwathan What are the factors you’re considering in the placement/order of `hover` in `group<label>-hover` and `group-hover/label`. Why in one is the label director after group but after the hover in the other? Would these two concepts have the same effect?

**Reply by ghost** [reply]
![Fcz7k4iWYAI7uR_](https://user-images.githubusercontent.com/113934633/194780232-b6817161-2275-4c2e-8ea6-ed39982d66e9.jpeg) would be nifty if it didn't need the group prefix at all.. like magically it looks for anything -hover and makes it a group. Would look alot cleaner. I wonder if you could see if the label element was a parent (group) or sibling (peer) and act accordingly when generating the css? its all voodoo to me haha

**Reply by olets** [reply]
What's the motivation for adding special syntax now that we have arbitrary variants?

`[.x:hover_&]:` is shorter than `group<x>-hover:`, `[.x:hover~&]` is shorter than `peer<x>-hover:`, and both are transparent and minimize the difference between what it means to learn Tailwind and what it means to learn CSS. (`<div x>` + `[[x]:hover_&]`/`[[x]:hover~&]` are even shorter, for cases where a class `x` would _only_ be added for use in a group or a peer.)

**Reply by RobinMalfait** [reply]
@olets a dedicated structure like `group` gives you better intellisense, but also the example you showed is technically incomplete the moment you start stacking your variants. 

For example if you use `group-hover:group-focus:underline` then the final result you expect is something like:
```css
.group:hover:focus .group-hover\:group-focus\:underline {}
```
Internally we use a `:merge(.group)` so that the states are merged together, if you don't do that OR in other words sue the syntax you shared then you would see this instead:
```css
.group:hover .group:focus .group-hover\:group-focus\:underline {}
```

So to correct for that you would have to introduce `:merge` here. Which looks like this:
```css
[:merge(.x):hover_%]:   // Your proposal
group<x>-hover:         // Current proposal
group/x-hover:          // New proposal
```

Hope that makes a bit of sense!

**Reply by royvanv** [reply]
> The square brackets isn't terrible but it just feels weird to use them again but for a different reason than we currently use them.

Square bracket are used for abritary selectors, so using them might be confusing and feel inconsistent.

> We've considered both `group(navitem)-hover:...` and `group[navitem]-hover:...` but not really excited about either option. 

Have you considered using curly brackets (`{`/`}`)?

**Reply by brandonmcconnell** [reply]
Using curly braces would result in a breaking change for many JS frameworks that already have their own usage of curly braces within classname strings. In other words, Tailwind devs everywhere would have to do something wonky like this:

```svelte
<SomeComponent class={`group{groupName}`} />
```

…as this would no longer work:

```svelte
<SomeComponent class="group{groupName}" /> <-- tries to interpolate a variable `groupName` into the string via JS
```

**Reply by AmProsius** [reply]
> So to correct for that you would have to introduce `:merge` here. Which looks like this:
> 
> ```css
> [:merge(.x):hover_%]:   // Your proposal
> group<x>-hover:         // Current proposal
> group/x-hover:          // New proposal
> ```
> 
> Hope that makes a bit of sense!

The new proposal feels a lot cleaner to me. How would you define the group then? Like `class="group/x"`?
Also what about `group-hover/x`?

**Reply by RobinMalfait** [reply]
@brandonmcconnell thats not entirely true. String interpolation only happens when using back ticks \` and if you use a `$` in front of the first `{`. So that should not be an issue. It looks a bit odd though.

**Reply by RobinMalfait** [reply]
@AmProsius sorry I use the wrong code in the example. We were still figuring things out there.

It would be like:
```html
<div class="group/foo">
  <div class="group-hover/foo:underline"></div>
</div>
```

**Reply by brandonmcconnell** [reply]
> @brandonmcconnell thats not entirely true. String interpolation only happens when using back ticks \` and if you use a `$` in front of the first `{`. So that should not be an issue. It looks a bit odd though.

@RobinMalfait At least in svelte, interpolation happens even when using regular quotes, e.g. `class="group/{group}"`

☝🏼 this will interpolate

**Reply by brandonmcconnell** [reply]
> @AmProsius sorry I use the wrong code in the example. We were still figuring things out there.
> 
> It would be like:
> ```html
> <div class="group/foo">
>   <div class="group-hover/foo:underline"></div>
> </div>
> ```

@RobinMalfait I'm very curious… Why `group-hover/name` and snot `group/name-hover`?

**Reply by adamwathan** [reply]
To be consistent with the modifier placement in classes like `bg-black/50` and to make it possible to use dashes in the modifier like `group-hover/main-nav` 👍🏻

**Comment by dtydev** [comment]
Hello
Have you planned any new features for the "Play CDN"?
For example, being able to isolate "style" with custom classes?
This would be appreciated for small projects.
Thanks for the job you do!

**Comment by czernika** [comment]
Some comments on `matchVariant`

1. I see editor complaining about 3rd argument as `matchVariant` in `config.d.ts` [has](https://github.com/tailwindlabs/tailwindcss/blob/master/types/config.d.ts#L299) 2 arguments

See this [TailwindPlay](https://play.tailwindcss.com/H0ZvxlFN3N?file=config)

2. In your example default values for "placement" dynamic variant

```js
{
    t: "top",
    r: "right",
    b: "bottom",
    l: "left",
}
```

should be under `values` key - it got me confused why it's not working :)

```js
{
    values: {
        t: "top",
        r: "right",
        b: "bottom",
        l: "left",
    }
}
```

**Comment by colinaut** [comment]
I like the use of / as it conceptually feels like system folders and
makes sense being a subset of the former

**Comment by wongjn** [comment]
Is it expected to get the following error with `npm` v8 when installing `insiders` with the core plugins?
```
> npm i -D tailwindcss@insiders @tailwindcss/typography

code ERESOLVE
ERESOLVE unable to resolve dependency tree

While resolving: undefined@undefined
Found: tailwindcss@0.0.0-insiders.c5ca373
node_modules/tailwindcss
  tailwindcss@"0.0.0-insiders.c5ca373" from the root project

Could not resolve dependency:
peer tailwindcss@">=3.0.0 || insiders" from @tailwindcss/typography@0.5.7
node_modules/@tailwindcss/typography
  @tailwindcss/typography@"*" from the root project
 ```

**Comment by jonnitto** [comment]
I have integrated it into a project, and everything runs smoothly. Especially the `<label>` feature is excellent and solves many problems. I have also already equipped a few internal plugins with `matchVariant`, everything runs without problems.

**Comment by VicGUTT** [comment]
Exciting stuff ! 👌

I see a nice opportunity for out-of-the-box support of `:has` here thanks to `matchVariant`. Has this been considered?

**Comment by davidhellmann** [comment]
Great! Like the Multiconfig feature a lot. Its great for special sites within a site like a Conference website. Thanks for that! 🔥

**Comment by michaelchiche** [comment]
is the nested group syntax available on the playground  yet? and using which format ? 
This is an amazing feature
